### PR TITLE
GORA-555: Improve Lucene query implementation with NumericRangeQuery

### PR DESCRIPTION
**Pull request by cuent**
Infer data type for building range queries. First, check if there is a numeric type if so cast to the appropriate type otherwise set to string datatype.

### CODE REVIEW

**Comment by lewismc** [comment]
This looks good @cuent thank you for taking the issue on. Please augment your pull request with Unit tests. Thank you

**Comment by djkevincr** [comment]
+1 @cuent This look certainly great. @lewismc mentioned you may add Lucene store specific tests. ( in TestLuceneStore.java. )

**Comment by cuent** [comment]
I was trying to add that but it creates two examples, employees and website, which have string keys. How can I extend that for creating a key with a numerical type? I was looking for that but I didn't found.

**Comment by lewismc** [comment]
@cuent you can easily implement individual tests for the module and use the `@Override` convention to implement the specific keys you need. See [TestHBaseStore.java](https://github.com/apache/gora/blob/master/gora-hbase/src/test/java/org/apache/gora/hbase/store/TestHBaseStore.java) for more details.

**Comment by cuent** [comment]
Yes, I reviewed the `DataStoreTestBase`, but it creates a `employeeStore` and a `webPageStore` which both have String kes. So even if I override the method, Query will ask for a String.

**Comment by lewismc** [comment]
So why don't you just create individual tests such as https://github.com/apache/gora/blob/master/gora-accumulo/src/test/java/org/apache/gora/accumulo/store/PartitionTest.java

**Comment by cuent** [comment]
I implemented the test. I also notice that in some places some keys are converted to `String` by default. I solved another issue by inferring the type while creating the document. It is working and passing all the tests.

File: gora-lucene/src/test/java/org/apache/gora/lucene/store/EmployeeInt.java [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+ *
+ *Unless required by applicable law or agreed to in writing, software
+ *distributed under the License is distributed on an "AS IS" BASIS,
+ *WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *See the License for the specific language governing permissions and
+ *limitations under the License.
+ */
+package org.apache.gora.lucene.store;
+
+public class EmployeeInt extends org.apache.gora.persistency.impl.PersistentBase implements org.apache.avro.specific.SpecificRecord, org.apache.gora.persistency.Persistent {
```
Comment (djkevincr) [inline gora-lucene/src/test/java/org/apache/gora/lucene/store/EmployeeInt.java]: Did you commit AVRO schema ( EmployeeInt.json ) for this EmployeeInt? Just to make sure, always use GoraCompiler to convert EmployeeInt.json --> EmployeeInt.Java.

File: gora-lucene/src/test/java/org/apache/gora/lucene/store/TestLuceneStore.java [hunk 2]
```diff
@@ -18,32 +18,40 @@
...
- * TestLuceneStore class executes tests for {@link org.apache.gora.lucene.store.LuceneStore}
+ * TestLuceneStore class executes tests for
+ * {@link org.apache.gora.lucene.store.LuceneStore}
  */
 public class TestLuceneStore extends DataStoreTestBase {
 
   private static final Logger LOG = LoggerFactory.getLogger(DataStoreTestUtil.class);
+  private DataStore<Integer, EmployeeInt> employeeIntStore;
+
+  @Before
```
Comment (djkevincr) [inline gora-lucene/src/test/java/org/apache/gora/lucene/store/TestLuceneStore.java]: @Before -  means this config() method is executed for each tests. Do we really need that behavior? It is only required to be executed before - testInferDataType() test case.

**Comment by djkevincr** [comment]
@cuent Thanks for considering my suggestions over the discussion we had on ticket. Noticed you have added changes that needed for put method in order use native types opposed to converting to Strings. This     is Awesome. Please address my comments, I will go ahead and locally test your PR.

File: gora-lucene/src/main/java/org/apache/gora/lucene/store/LuceneStore.java [hunk 3]
```diff
@@ -545,7 +546,17 @@ public void put(K key, T persistent) {
     }
     LOG.info("DOCUMENT: {}", doc);
     try {
-      doc.add(new StringField(mapping.getPrimaryKey(), key.toString(), Store.YES));
+      if (key instanceof Integer) {
+        doc.add(new IntPoint(mapping.getPrimaryKey(), ((Integer) key).intValue()));
```
Comment (djkevincr) [inline gora-lucene/src/main/java/org/apache/gora/lucene/store/LuceneStore.java]: You could use auto unboxing -  wrapper class to the primitive class
new IntPoint(mapping.getPrimaryKey(), ((Integer) key).intValue()) ===
new IntPoint(mapping.getPrimaryKey(), (Integer) key)

File: gora-lucene/src/main/java/org/apache/gora/lucene/query/LuceneQuery.java [hunk 4]
```diff
@@ -51,22 +57,41 @@ public Query toLuceneQuery() {
...
+        q = inferType(pk, getStartKey(), getEndKey());
       }
     }
     return q;
   }
+
+  private <T> Query inferType(String pk, T lower, T upper) {
+    if (((lower != null && lower.getClass() == Integer.class)
+            || (upper != null && upper.getClass() == Integer.class))) {
+      int ilower = lower == null ? Integer.MIN_VALUE : Integer.valueOf(lower.toString());
```
Comment (djkevincr) [inline gora-lucene/src/main/java/org/apache/gora/lucene/query/LuceneQuery.java]: int ilower = lower == null ? Integer.MIN_VALUE : Integer.valueOf(lower.toString());
If the value here is in Integer class ( lower.getClass() == Integer.class )
Is it necessary to convert to String and take the Integer value using Integer.valueOf(lower.toString())  === 
int ilower = lower == null ? Integer.MIN_VALUE : (Integer) lower;

This is the same for other cases in below for float, long, double.

File: gora-lucene/src/main/java/org/apache/gora/lucene/query/LuceneQuery.java [hunk 5]
```diff
@@ -51,22 +57,41 @@ public Query toLuceneQuery() {
...
+        return new MatchAllDocsQuery();
       } else {
-        q = TermRangeQuery.newStringRange(pk, lower, upper, true, true);
+        q = inferType(pk, getStartKey(), getEndKey());
       }
     }
     return q;
   }
+
+  private <T> Query inferType(String pk, T lower, T upper) {
```
Comment (djkevincr) [inline gora-lucene/src/main/java/org/apache/gora/lucene/query/LuceneQuery.java]: Don't you think it s more readable when we use Key type - K directly here as - 
private Query inferType(String pk, K lower, K upper) {

**Comment by cuent** [comment]
@djkevincr done!

**Comment by djkevincr** [comment]
@cuent Thank you for addressing my comments. Were you able add the EmployeeInt.json AVRO schema? This should be added to gora-core/src/examples/avro directory. You will see - employee.json
You will have to rename the name and type for ssn. Please make sure you regenerate using the json file. This looks Awesome and ready to be merged :)

File: gora-lucene/src/test/java/org/apache/gora/lucene/store/EmployeeInt.java (hunk 1 above)
Comment (djkevincr) [reply gora-lucene/src/test/java/org/apache/gora/lucene/store/EmployeeInt.java]: Please add EmployeeInt.json file separately to gora-core module.

Comment (cuent) [reply gora-lucene/src/test/java/org/apache/gora/lucene/store/EmployeeInt.java]: it was added, forgot to push the commit 😄

**Comment by djkevincr** [comment]
I have approved the PR, will locally test and merge to the master.

**Comment by djkevincr** [comment]
@lewismc Please raise if you do have any concerns. Otherwise will proceed to merge.

**Pull request merged**